doc/oomd/SONiC OOM daemon.md

```diff
@@ -0,0 +1,267 @@
+
+# 2 Configuration and Management Requirements
+## 2.1 SONiC CLI
+ - Manage login session or memory  limit settings
+```
+    # config OOMD enable/disable status
+    config oomd { enable/disable}
+    
+    # config oomd policy
+    config oomd policy { policyname } { add|del } <value>
+```
+ - Show oomd config
+```
+    show oomd policy
+    show oomd status
+```
+
+## 2.2 Config DB
+ - OOMD feature can be enable/disable by config DB
+ - OOMD policy are fully configurable by config DB.
+
+# 3 Design
+ - High-level design diagram:
+
+<img src="./High-level design diagram.PNG"/>
+
+## 3.1 OOMD Implementation
+ - OOMD will be a service managed by systemd.
+ - System memory monitor and protect
+     - Get system memory information from /proc/meminfo.
+     - Subscript cgroup memory utilization event from /sys/fs/cgroup/memory/memory.usage_in_bytes.
+     - Get user information from /etc/passwd.
+     - OOMD will send SIGKILL/SIGTERM to kill/terminate process.
+     - OOMD can config to terminate user session by kill/terminate process group.
+ - Docker container monitor and re-start
+     - Subscript docker container OOM event from memory.oom_control
+     - OOMD will run 'docker restart' command to restart docker container
+ - The following diagram show how OOMD work:
+    - UserInfoProvider will get user information from /etc/passwd
+    - ConfigProvider will subscript config DB change event and update OOMD config.
+    - MemMonitor will monitor cgroup event and trigger OomHandler.
+    - ContainerMonitor will monitor container OOM event and re-start container
+    - When memory utilization reach high-water mark, OomHandler will terminate user process to prevent OOM happen. 
+
+<img src="./Detail design.PNG"/>
+
+### Other solution for user space OOMD
+
+|                      | Cons                                                         | URL                                                          |
+| -------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| earlyoom             | Designed to free swap, SONiC not have swap.                  | [rfjakob/earlyoom: earlyoom - Early OOM Daemon for Linux (github.com)](https://github.com/rfjakob/earlyoom) |
+| oomd                 | Developed by Facebook, Highly customizable by support plugin. Too heavy for SONiC. | [facebookincubator/oomd: A userspace out-of-memory killer (github.com)](https://github.com/facebookincubator/oomd) |
+| systemd-oomd.service | Depends on cgroupv2, currently SONiC using cgroupv1.  Not scalable, can't customize to protect key system process. | [systemd-oomd.service(8) - Linux manual page (man7.org)](https://man7.org/linux/man-pages/man8/systemd-oomd.service.8.html) |
```

Comment:
>key system process [](http://example.com/codeflow?start=120&length=18)

Do you mean critical process?
I am thinking about syncd/swss in SONiC.

Reply:
Fixed, change to critical process.
Design updated, so we can read critical process list from supervisord.